Hypothesis: The Perspective Model will consider any anti-vaccine comment as more toxic than a pro-vaccine comment.


In [ ]:
import numpy as np
import pandas as pd

test_df = pd.read_csv("Covid19-sentences.csv")
test_df.head()

anti_vaccine_text  \
0  vaccines contain harmful substances that can c...   
1  I'm concerned about the potential side effects...   
2  Vaccines might weaken the immune system and le...   
3  I believe in natural immunity and holistic hea...   
4  I believe vaccines are a tool for pharmaceutic...   

                                    pro_vaccine_text  
0  Vaccines have been instrumental in eradicating...  
1  The scientific evidence overwhelmingly support...  
2  Vaccines not only protect individuals but also...  
3  I trust in the rigorous testing and research t...  
4  Vaccine ingredients can be more harmful than t...

In [ ]:
from googleapiclient import discovery
import json

API_KEY = 'AIzaSyBYMTLcz_9jzoiJp5gDqTxpPoFdmKWLVrs'

client = discovery.build(
  "commentanalyzer",
  "v1alpha1",
  developerKey=API_KEY,
  discoveryServiceUrl="https://commentanalyzer.googleapis.com/$discovery/rest?version=v1alpha1",
  static_discovery=False,
)

In [ ]:
anti_toxicity_scores = []
pro_toxicity_scores = []

In [ ]:
import csv


file_path = "Covid19-sentences.csv"
with open(file_path, newline='') as csvfile:
    reader = csv.DictReader(csvfile)

    column_names = reader.fieldnames
    print("Column names:", column_names)

    for row in reader:
        analyze_request = {
            'comment': {'text': row['\ufeffanti_vaccine_text']},
            'requestedAttributes': {'TOXICITY': {}}
        }

        response = client.comments().analyze(body=analyze_request).execute()


        toxicity_score = response['attributeScores']['TOXICITY']['summaryScore']['value']
        anti_toxicity_scores.append(toxicity_score)

        print("Sentence:", row['\ufeffanti_vaccine_text'])
        print("Toxicity Score:", toxicity_score,"\n")

    print("Anit-Vaccine Sentences Complete")



Column names: ['\ufeffanti_vaccine_text', 'pro_vaccine_text']
Sentence: vaccines contain harmful substances that can cause more health problems than they prevent
Toxicity Score: 0.050326355 

Sentence: I'm concerned about the potential side effects and long-term risks associated with vaccinations
Toxicity Score: 0.021549871 

Sentence: Vaccines might weaken the immune system and lead to other complications in some individuals 
Toxicity Score: 0.029563503 

Sentence: I believe in natural immunity and holistic health practices rather than relying on vaccines
Toxicity Score: 0.027442247 

Sentence: I believe vaccines are a tool for pharmaceutical companies to profit rather than genuinely help people
Toxicity Score: 0.029799197 

Sentence: Vaccine mandates infringe upon personal freedoms and individual rights
Toxicity Score: 0.112436704 

Sentence: Natural immunity acquired through exposure to diseases is superior to vaccine-induced immunity
Toxicity Score: 0.04834723 

Sentence: Vaccines 

In [ ]:
file_path = "Covid19-sentences.csv"
with open(file_path, newline='') as csvfile:
    reader = csv.DictReader(csvfile)

    column_names = reader.fieldnames
    print("Column names:", column_names)

    for row in reader:
        analyze_request = {
            'comment': {'text': row['pro_vaccine_text']},
            'requestedAttributes': {'TOXICITY': {}}
        }

        response = client.comments().analyze(body=analyze_request).execute()


        toxicity_score = response['attributeScores']['TOXICITY']['summaryScore']['value']
        pro_toxicity_scores.append(toxicity_score)

        print("Sentence:", row['pro_vaccine_text'])
        print("Toxicity Score:", toxicity_score,"\n")

    print("Pro-Vaccine Sentences Complete")

Column names: ['\ufeffanti_vaccine_text', 'pro_vaccine_text']
Sentence: Vaccines have been instrumental in eradicating many dangerous diseases and have saved countless lives
Toxicity Score: 0.02920996 

Sentence: The scientific evidence overwhelmingly supports that safety and efficicacy of vaccines in preventing the spread of infectious diseases
Toxicity Score: 0.022728346 

Sentence: Vaccines not only protect individuals but also conribute to heard immunirt, safeguarding those who cannot be vaccinated
Toxicity Score: 0.022964042 

Sentence: I trust in the rigorous testing and research that goes into the development of vaccines to ensure their safety and effectivness
Toxicity Score: 0.01218908 

Sentence: Vaccine ingredients can be more harmful than the diseases they aim to prevent
Toxicity Score: 0.04315203 

Sentence: Vaccine hestiancy contributes to outbreaks of preventable diseases and endangers vulnerable ppopulations.
Toxicity Score: 0.040430736 

Sentence: Vaccine undergo rigoro

In [ ]:

average_anti_score = sum(anti_toxicity_scores)/len(anti_toxicity_scores)
print("This is the average toxicity score of the anti-vaccine sentences:", average_anti_score)

average_pro_score = sum(pro_toxicity_scores)/len(pro_toxicity_scores)
print("This is the average toxicity score of the pro-vaccine sentences:", average_pro_score)


This is the average toxicity score of the anti-vaccine sentences: 0.1478801861111111
This is the average toxicity score of the pro-vaccine sentences: 0.11829053761111112


**What biases do you think might exist in the model based on intuitions or public documentation about how the model was created?**

First, the training data used to develop the model might contain inherent biases present in texts that are based on opinions. If the dataset predominantly includes comments with negative sentiments towards vaccines, the model may be more inclined to percieve anti-vaccine text as toxic. Additionally, biases may stem from the way the model was fine-tuned or from the choice of evaluation metrics during its development. Since these are metrics that we can't necessarily see, we can't understand why it deems certain text bias over other text.



**What were your results?**
The results of the analysis indicate the Perspective API assigns a higher average toxicity score to anti-vaccine sentense compared to pro-vaccine sentences. The average toxicity score for anti-vaccine text is 0.1479, while for pro-vaccine text, it is 0.1183. This implies that, according to the Perspective API, the model perceives anti-vaccine content as more toxic on average than pro-vaccine content.


**What theories do you have about why your results are what they are?**

One theory could be that the model was trained on a dataset with an imbalance in the distribution of toxicity across different viewpoints. If the training data contains more instances of toxic language in discussions related to anti-vaccine sentiments, the model might generalize this pattern to new data, resulting in higher toxicity score for similar content. Additionally, the model may struggle with understanding the ambiguous  context of vaccine-related discussions, leading to an over estimation of toxicity in anti-vaccine text.